rust/nasl-builtin-std/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    builder.push_register(nasl_builtin_raw_ip::RawIp)
+}
+
+#[cfg(feature = "nasl-builtin-ssh")]
+fn add_ssh<K: AsRef<str>>(
+    builder: nasl_builtin_utils::NaslfunctionRegisterBuilder<K>,
+) -> nasl_builtin_utils::NaslfunctionRegisterBuilder<K> {
+    builder.push_register(nasl_builtin_ssh::Ssh::default())
+}
+
+/// Contains the key as well as the dispatcher.
+///
+/// This is to ensure that the key and the dispatcher do have the same type.
+/// It makes it also a bit easier to use as the key as well as the dispatcher are usually created
+/// together.
+pub struct KeyDispatcherSet<K: AsRef<str>> {
+    key: K,
+    storage: Box<dyn Storage<K>>,
+}
+
+/// The context builder.
+///
+/// This is the main entry point for the nasl interpreter and adds all the functions defined in
+/// [nasl_std_functions] to functions register.
+pub struct ContextBuilder<K: AsRef<str>, S> {
+    /// The key and dispatcher set.
+    pub key: S,
+    /// The target to test.
+    pub target: String,
+    /// The loader to load the nasl files.
+    pub loader: Box<dyn Loader>,
+    /// The logger to log.
+    pub logger: Box<dyn NaslLogger>,
+    /// The functions available to the nasl script.
+    pub functions: NaslFunctionRegister<K>,
+}
+
+impl Default for ContextBuilder<String, KeyDispatcherSet<String>> {
+    fn default() -> Self {
+        Self {
+            key: KeyDispatcherSet {
+                key: Default::default(),
+                storage: Box::<DefaultDispatcher<String>>::default(),
+            },
+            target: Default::default(),
+            loader: Default::default(),
+            logger: Default::default(),
+            functions: nasl_std_functions(),
+        }
+    }
+}
+
+impl<K: AsRef<str>> ContextBuilder<K, KeyDispatcherSet<K>> {
+    /// Creates a new context builder with the given key and storage.
+    pub fn new(key: K, storage: Box<dyn Storage<K>>) -> Self {
+        Self {
+            key: KeyDispatcherSet { key, storage },
+            target: Default::default(),
+            loader: Default::default(),
+            logger: Default::default(),
+            functions: nasl_std_functions(),
+        }
+    }
+}
+
+impl<K: AsRef<str>, S> ContextBuilder<K, S> {
+    /// Sets the target to test.
+    pub fn target(mut self, target: String) -> Self {
+        self.target = target;
+        self
+    }
+
+    /// Sets the loader to load the nasl files.
+    pub fn loader<L: Loader + 'static>(mut self, loader: L) -> Self {
+        self.loader = Box::new(loader);
+        self
+    }
+
+    /// Sets the logger to log.
+    pub fn logger<L: NaslLogger + 'static>(mut self, logger: L) -> Self {
+        self.logger = Box::new(logger);
+        self
+    }
+
+    /// Sets the functions available to the nasl script.
+    pub fn functions(mut self, functions: NaslFunctionRegister<K>) -> Self {
+        self.functions = functions;
+        self
+    }
+}
+
+impl<K: AsRef<str>> ContextBuilder<K, KeyDispatcherSet<K>> {
```

Comment:
This implementation is already done for `new()` above in the line 151

Reply:
Can you elaborate it a bit? 

The new function in line 151 creates a new ContextBuilder while the build function marked here creates a Context. So those are not the same and do completely different things unless I overlooked something.

Reply:
yes, build() and new() do different things, but both are implementation for `ContextBuilder` What I mean is that they can be inside the same `impl {}` block. Just for easier reading.

Reply:
ah, got it. You're right thanks.